ktor-server/ktor-server-plugins/ktor-server-single-page/jvm/src/io/ktor/plugins/SinglePageApplication.kt

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2014-2021 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.plugins
+
+import io.ktor.http.*
+import io.ktor.server.application.*
+import io.ktor.server.http.content.*
+import io.ktor.server.request.*
+import io.ktor.server.response.*
+import io.ktor.server.routing.*
+import io.ktor.util.*
+import io.ktor.util.pipeline.*
+import java.io.*
+
+/**
+ * A plugin that allows you to serve a single-page application
+ *
+ * A basic plugin configuration for the application served from the filesPath folder
+ * with index.html as a default file:
+ *
+ * install(SinglePageApplication) {
+ *   filesPath = "application/project_path"
+ * }
+ */
+public class SinglePageApplication internal constructor(configuration: Configuration) {
+
+    internal val defaultPage: String = configuration.defaultPage
+
+    internal val applicationRoute: String = configuration.applicationRoute
+
+    internal val filesPath: String = configuration.filesPath
+
+    internal val ignoredFiles: MutableList<(String) -> Boolean> = configuration.ignoredFiles
+
+    internal val usePackageNames: Boolean = configuration.useResources
+
+    /**
+     * Configuration for the [SinglePageApplication] plugin
+     */
+    public class Configuration(
+        /**
+         * The default name of a file or resource to serve when [applicationRoute] is requested
+         */
+        public var defaultPage: String = "index.html",
+
+        /**
+         * The URL path under which the content should be served
+         */
+        public var applicationRoute: String = "/",
+
+        /**
+         * The path under which the static content is located.
+         * Corresponds to the folder path if the [useResources] is false, resource path otherwise
+         */
+        public var filesPath: String = "",
+
+        /**
+         * Specifies if static content is a resource or folder
+         */
+        public var useResources: Boolean = false,
+
+        /**
+         * A list of callbacks checking if a file or resource in [filesPath] is ignored.
+         * Requests for such files or resources fails with the 404 Forbidden status
+         */
+        internal val ignoredFiles: MutableList<(String) -> Boolean> = mutableListOf()
+    ) {
+        /**
+         * Registers a [block] in [ignoredFiles]
+         * [block] returns true if [path] should be ignored.
+         */
+        public fun ignoreFiles(block: (path: String) -> Boolean) {
+            ignoredFiles += block
+        }
+
+        /**
+         * Creates an application configuration for the Angular project.
+         * Resources will be shared from the filesPath directory. The root file is index.html
+         */
+        public fun angular(filesPath: String) {
```

Comment:
This and below can be extension functions.
Is this enough configurations?

Reply:
Functions are extentions now. 
This configuration is basic: use index.html, share files from `filesPath`, no ignore files. As far as I understand, every SPA framework from this functions works with this configuration by default. Maybe there is something more I don't know about?